interfaces/kepware/config.html

```diff
@@ -357,206 +369,335 @@
+              });
+            });
+          });
+          return results;
+        }
+        
+        const tagListUI = document.createElement('div');
+        const tagListUISeparator = document.createElement('div');
+        tagListUISeparator.style.border = '2px dashed white';
+        tagListUI.appendChild(tagListUISeparator);
+        realityServer.domObjects.appendChild(tagListUI);
+        
+        if (machines.length === 0) {
+          const connectMessage = document.createElement('p');
+          connectMessage.innerText = 'Fill in the details above and refresh the page to see available tags. Please ensure that you have entered the correct credentials and url.'
+          connectMessage.classList.add('yellow');
+          connectMessage.style.marginTop = '20px';
+          realityServer.domObjects.appendChild(connectMessage);
+        } else {
+          const tagListUIHeading = document.createElement('h2');
+          tagListUIHeading.innerText = 'Available Tags';
+          tagListUIHeading.style.marginBottom = '0px';
+          tagListUI.appendChild(tagListUIHeading);
+          
+          const buttonDiv = document.createElement('div');
+          buttonDiv.style.marginBottom = '10px';
+          const resetButton = document.createElement('button');
+          resetButton.innerText = "Reset Changes";
+          resetButton.classList.add('reset', 'item', 'button', 'red', 'clickAble');
+          resetButton.addEventListener('click', () => {
+            realityServer.render();
+          });
+          buttonDiv.appendChild(resetButton);
+          const saveButton = document.createElement('button');
+          saveButton.innerText = "Save Changes";
+          saveButton.classList.add('save', 'item', 'button', 'green', 'clickAble');
+          saveButton.addEventListener('click', () => {
+            realityServer.hardwareInterface.settings['enabledTags'].value = getSelectedObjects(machines);
+            realityServer.writeHardwareInterfaceSettings(settings, ['enabledTags'], function(res) {
+              console.log('got response... re-render');
+              realityServer.render();
+            });
+          });
+          buttonDiv.appendChild(saveButton);
+          tagListUI.appendChild(buttonDiv);
+          
+          const searchLabel = document.createElement('span');
+          searchLabel.innerText = 'Filter results: '
+          tagListUI.appendChild(searchLabel);
+          const searchBox = document.createElement('input');
+          searchBox.style.marginBottom = '10px';
+          searchBox.addEventListener('input', e => {
+            const searchText = e.target.value;
+            machines.forEach(machine => {
+              if (!machine.name.toLowerCase().includes(searchText.toLowerCase()) && !machine.frames.some(frame => frame.name.toLowerCase().includes(searchText.toLowerCase()) || frame.tags.some(tag => tag.name.toLowerCase().includes(searchText.toLowerCase())))) {
```

Comment:
nit: this line should be wrapped and/or refactored out to individual variables (ideally with self-documenting names)